src/domain/login/LoginViewModel.js

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2020 Bruno Windels <bruno@windels.cloud>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {ViewModel} from "../ViewModel.js";
+import {PasswordLoginViewModel} from "./PasswordLoginViewModel.js";
+import {SSOLoginViewModel} from "./SSOLoginViewModel.js";
+import {normalizeHomeserver} from "./common.js";
+
+export class LoginViewModel extends ViewModel {
+    constructor(options) {
+        super(options);
+        const {ready, defaultHomeServer, createSessionContainer, loginToken} = options;
+        this._createSessionContainer = createSessionContainer;
+        this._ready = ready;
+        this._defaultHomeServer = defaultHomeServer;
+        this._loginToken = loginToken;
+        this._sessionContainer = this._createSessionContainer();
+        this._loginOptions = null;
+        this._start();
+    }
+
+    get passwordLoginViewModel() { return this._passwordLoginViewModel; }
+    get ssoLoginViewModel() { return this._ssoLoginViewModel; }
+    get loadViewModel() {return this._loadViewModel; }
+
+    async _start() {
+        if (this._loginToken) {
+            this._ssoLoginViewModel = this.track(new SSOLoginViewModel(this.childOptions({loginToken: this._loginToken})));
```

Comment:
no dynamic properties, please. Ideally declare properties in the constructor, or in a method called by the constructor, but not conditionally like here.

Reply:
Here, you'd actually have two members: one for StartSSOLoginViewModel, and one for CompleteSSOLoginViewModel